src/utils/scheduler.h

```diff
@@ -299,6 +301,219 @@ class Watchdog : private impl::Notifiable {
 	}
 };
 
+/**
+ * @brief An abstract class that can be overridden to run long-running tasks and encapsulate
+ * task-related data.
+ *
+ * Client code should use this class by deriving from it and overriding AsyncTask::task.
+ *
+ * For simpler tasks that just require a function to be run periodically, consider
+ * PeriodicTask.
+ *
+ * @tparam Clock The clock to use for timing.
+ */
+template <typename Clock = std::chrono::steady_clock>
+class AsyncTask : private virtual impl::Notifiable {
+public:
+	/**
+	 * @brief Construct a new task.
+	 *
+	 * @param name The name of this task, for logging purposes.
+	 */
+	AsyncTask(const std::optional<std::string>& name = std::nullopt)
+		: name(name), running(false), quitting(false) {}
+
+	AsyncTask(const AsyncTask&) = delete;
+
+	virtual ~AsyncTask() {
+		stop();
+		if (thread.joinable()) {
+			thread.join();
+		}
```

Comment:
Is this necessary? `stop()` joins the thread and waits for the task to finish if it's running, and if it's not running, there's no need to join and wait, if my understanding is correct. Is this just for extra safety?

Reply:
Good question, `stop()` only joins the thread if `running` is true, but there is the edge case where the thread has already exited. In that case, `running` is false, but the thread still needs to be joined to clean up its resources.

`running` tracks whether the task is logically running or not, not necessarily if the thread is active or not. So, in the destructor, we need to explicitly join the thread to make sure we clean it up.